persistent-postgresql/Database/Persist/Postgresql/Internal.hs

```diff
@@ -372,29 +390,71 @@ data AlterDB
 -- current state in the database to the state described in
 -- Haskell.
 --
--- @since 2.17.0.0
-mockMigrateStructured
+-- @since 2.17.1.0
+migrateStructured
     :: [EntityDef]
+    -> (Text -> IO Statement)
     -> EntityDef
     -> IO (Either [Text] [AlterDB])
-mockMigrateStructured allDefs entity = do
-    case partitionEithers [] of
-        ([], old'') -> return $ Right $ migrationText False old''
+migrateStructured allDefs getter entity = do
+    old <- getColumns getter entity newcols'
+    case partitionEithers old of
+        ([], old'') -> do
+            exists' <-
+                if null old
+                    then doesTableExist getter name
+                    else return True
+            return $ Right $ migrationText exists' old''
         (errs, _) -> return $ Left errs
   where
     name = getEntityDBName entity
-    migrationText exists' old'' =
-        if not exists'
-            then createText newcols fdefs udspair
-            else
-                let
-                    (acs, ats) = getAlters allDefs entity (newcols, udspair) old'
-                    acs' = map (AlterColumn name) acs
-                    ats' = map (AlterTable name) ats
-                 in
-                    acs' ++ ats'
+    (newcols', udefs, fdefs) = postgresMkColumns allDefs entity
+    migrationText exists' old''
+        | not exists' =
+            createText newcols fdefs udspair
+        | otherwise =
+            let
+                (acs, ats) =
+                    getAlters allDefs entity (newcols, udspair) old'
+                acs' = map (AlterColumn name) acs
+                ats' = map (AlterTable name) ats
+             in
+                acs' ++ ats'
       where
         old' = partitionEithers old''
+        newcols = filter (not . safeToRemove entity . cName) newcols'
+        udspair = map udToPair udefs
+    -- Check for table existence if there are no columns, workaround
+    -- for https://github.com/yesodweb/persistent/issues/152
+
+    createText newcols fdefs_ udspair =
+        (addTable newcols entity) : uniques ++ references ++ foreignsAlt
+      where
+        uniques = flip concatMap udspair $ \(uname, ucols) ->
+            [AlterTable name $ AddUniqueConstraint uname ucols]
+        references =
+            mapMaybe
+                ( \Column{cName, cReference} ->
+                    getAddReference allDefs entity cName =<< cReference
+                )
+                newcols
+        foreignsAlt = mapMaybe (mkForeignAlt entity) fdefs_
+
+-- | Returns a structured representation of all of the
+-- DB changes required to migrate the Entity to the state
+-- described in Haskell, assuming it currently does not
+-- exist in the database.
+--
+-- @since 2.17.1.0
+mockMigrateStructured
+    :: [EntityDef]
+    -> EntityDef
+    -> IO (Either [Text] [AlterDB])
+mockMigrateStructured allDefs entity = return $ Right migrationText
```

Comment:
This feels odd to me - if we're always succeeding with `Right`, why not just do that? Same re `IO`. Is this suppose dto be doing more?